// react-native-autocomplete-dropdown/lib/commonjs/HOC/withFadeAnimation.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.withFadeAnimation = void 0;
var _react = _interopRequireWildcard(require("react"));
var _reactNative = require("react-native");
var _jsxRuntime = require("react/jsx-runtime");
function _interopRequireWildcard(e, t) { if ("function" == typeof WeakMap) var r = new WeakMap(), n = new WeakMap(); return (_interopRequireWildcard = function (e, t) { if (!t && e && e.__esModule) return e; var o, i, f = { __proto__: null, default: e }; if (null === e || "object" != typeof e && "function" != typeof e) return f; if (o = t ? n : r) { if (o.has(e)) return o.get(e); o.set(e, f); } for (const t in e) "default" !== t && {}.hasOwnProperty.call(e, t) && ((i = (o = Object.defineProperty) && Object.getOwnPropertyDescriptor(e, t)) && (i.get || i.set) ? o(f, t, i) : f[t] = e[t]); return f; })(e, t); }
const withFadeAnimation = (WrappedComponent, {
  containerStyle
} = {}) => {
  return props => {
    const opacityAnimationValue = (0, _react.useRef)(new _reactNative.Animated.Value(0)).current;
    (0, _react.useEffect)(() => {
      _reactNative.Animated.timing(opacityAnimationValue, {
        duration: 800,
        toValue: 1,
        useNativeDriver: true,
        easing: _reactNative.Easing.bezier(0.3, 0.58, 0.25, 0.99)
      }).start();
    }, [opacityAnimationValue]);
    return /*#__PURE__*/(0, _jsxRuntime.jsx)(_reactNative.Animated.View, {
      style: [containerStyle, {
        opacity: opacityAnimationValue
      }],
      children: /*#__PURE__*/(0, _jsxRuntime.jsx)(WrappedComponent, {
        ...props
      })
    });
  };
};
exports.withFadeAnimation = withFadeAnimation;
